### ISSUE TITLE
Disable specific rules from auto-fixing on save?

### ISSUE THREAD

**Issue by foucdeg** [astral-sh/ruff-vscode#260]
When coding, it is annoying that our temporary imports and variables that we haven't _yet_ used are removed every time we save the file.

We'd like the Ruff extension to auto-fix on save in general (and we have that activated through), but not F401 (unused imports) and F841 (unused variables). 

We do want the rule to run autofix in other contexts (pre-commit hook, CI). Is there a way to do that?

**Comment by phillies** [comment]
I totally agree. Especially once https://github.com/astral-sh/ruff-vscode/issues/247 is fixed the automatic fix of F401 and F841 might lead to confusion and annoyance. I guess we would need 2 configuration sets, one for creating the warnings/highlights and one for fixing.

I would propose maybe something like this
```
    "ruff.args": [
        "--line-length=97",
        "--select=C,E,F,W,B,E501,I001",
        "--extend-ignore=E203"
    ],
    "ruff.fixAll.args": [
        "--line-length=97",
        "--select=C,E,F,W,B,E501,I001",
        "--extend-ignore=E203,F841,F401"
    ],
```

**Comment by francesco086** [comment]
I'm also interested as F401 drives me nut. I had to disable it altogether for now.

**Comment by albertotb** [comment]
This feature would be awesome, I have to keep re-typing imports whenever I comment out parts of the code while developing.

**Comment by charliermarsh** [comment]
Totally hear you. I should note that you can disable the fix for F401 entirely via:
```json
    "ruff.args": [
        "--unfixable=F401"
    ],
```

This'll prevent Ruff from removing unused imports on save. But it'll also prevent Ruff from removing unused imports when you run "Fix all" and such, so it's not a perfect solution.

**Comment by aneeshusa** [comment]
Just discussed this with Charlie live at PyCon - we currently have `"ruff.lint.args": ["--ignore=F401"]` in VSCode workspace settings to set the default for our org, but not ruff.toml so this is still enforced in CI. We would like to be able to "opt-in" to removing unused imports in VSCode, some ideas we discussed:
- VSCode user settings don't work as workspace settings [take precedence](https://code.visualstudio.com/docs/getstarted/settings#_settings-precedence)
- Having an explicit VSCode action for this - either as part of Organize Imports or a standalone "Remove unused imports" action
- Making F401 only run when saving a file, but not while just editing (not sure if this is too aggressive)

**Comment by kaddkaka** [comment]
Wouldn't a regular lsp code action be preferable as that would be useful for other editors as well?

**Comment by T-256** [comment]
> * VSCode user settings don't work as workspace settings [take precedence](https://code.visualstudio.com/docs/getstarted/settings#_settings-precedence)

@aneeshusa Does #457 solve it? you can set new `ruff.configurationPreference` in your VSCode settings.

**Comment by Cito** [comment]
Note that in newer versions of ruff you need to set
```json
    "ruff.lint.args": [
        "--unfixable=F401"
    ],
```

**Comment by amaurygelin** [comment]
> Note that in newer versions of ruff you need to set
> 
> ```json
>     "ruff.lint.args": [
>         "--unfixable=F401"
>     ],
> ```

Hello, it does not work for me, unused imports are still deleted.
Here is my settings.json:

    "[python]": {
      "editor.formatOnSave": true,
      "editor.codeActionsOnSave": {
        "source.fixAll": "explicit",
        "source.organizeImports": "explicit"
      },
      "editor.defaultFormatter": "charliermarsh.ruff"
    },
    "ruff.lint.args": [
      "--unfixable=F401"
    ],

**Comment by dhruvmanila** [comment]
@amaurygelin The [`ruff.lint.args`](https://docs.astral.sh/ruff/editors/settings/#lintargs) setting isn't used by the native server (you've set `ruff.nativeServer` to `true`). For the native server, there are multiple options based on your use-case:
1. You can use a local configuration file (`pyproject.toml`) which will be picked up by Ruff
2. Use a global configuration file and define it in [`ruff.configuration`](https://docs.astral.sh/ruff/editors/settings/#configuration)

**Comment by T-256** [comment]
> We do want the rule to autofix in other contexts (pre-commit hook, CI). Is there a way to do that?



> 2\. Use a global configuration file and define it in [`ruff.configuration`](https://docs.astral.sh/ruff/editors/settings/#configuration)

Could we make it to accept json values?
It'd be easier for user than creating files manually and pass cross-platform aware paths.
```json
"ruff.nativeServer": true,
"ruff.configuration": {
   "unfixable": ["F401", "F801"]
}
```
we can either drop string value support for `ruff.configuration` in favor of `{"extend": "path/to/ruff.toml"}` value, Or we can separate this setting.

I believe this setting also helps to:
- #6
- #441 
- #450

**Comment by dhruvmanila** [comment]
> Could we make it to accept json values?
> It'd be easier for user than creating files manually and pass cross-platform aware paths.

We considered this for the configuration proposal but decided against it for now because:

1. We want to provide auto-completion, which requires including the entire schema in `package.json`.
2. Compatibility with different Ruff versions is a challenge. Shipping a schema tied to a specific extension version could limit flexibility since we allow users to use a specific Ruff version with `ruff.path`.

There's an option to allow any value to pass through to Ruff, leaving validation to the executable. However, this might result in a less optimal user experience due to the lack of auto-completion and delayed feedback on settings correctness.

**Comment by dhruvmanila** [comment]
I'm interested in supporting this and have created an issue with more details at https://github.com/astral-sh/ruff/issues/12709. If anyone wants to work on this, feel free to ping me / comment on the Ruff issue.

**Comment by alicederyn** [comment]
I don't think it's excessive to say this makes this plugin's auto-on-save features unusable 😢 I really hope this gets fixed.

**Comment by dhruvmanila** [comment]
> Does [#702](https://github.com/astral-sh/ruff-vscode/pull/702) solve this issue?

That's a good point. I think it can partially resolve it i.e., a user can configure Ruff in an editor context to mark certain rules as `unfixable` which will then be considered when applying code actions on save. But, it'll also mean that those won't show up in the lightbulb menu nor as a "Quick fix" in the diagnostic popup.

Consider this example config:
```json
{
  "ruff.configuration": {
    "lint": {
      "unfixable": ["F541"]
    }
  }
}
```

So, I think it's still useful to provide that feature so that the user still have the option to fix it if they want to.

> Do you mean Ruff Extension can send lint diagnostics for `settings.json`? Cool, it's better than nothing
> It can get diagnostics by validating passed value to `ruff.configuaretion` via `ruff config`.

No, I meant what we've implemented in https://github.com/astral-sh/ruff/pull/16296 which is that `ruff.configuration` will accept an `object` that represents the Ruff config which will then be passed onto the chosen `ruff` executable. The validation is performed on the server side which will notify the user if it fails.

> It needs investigation to check is there any other extensions that implemented auto-completion for settings.json. I think it makes sense at least provided a way from VSCode to let extension set auto-completion values dynamically during runtime instead of specifying statically in `package.json`.

Yeah, I don't know if that's even possible because the extension would need to provide suggestion based on the selected `ruff` executable which is might not even be known while the user is editing `settings.json`.

**Comment by dhruvmanila** [comment]
@T-256 Thank you for providing other options! I think any design discussions should be moved to https://github.com/astral-sh/ruff/issues/12709 instead.
